include/pybind11/detail/internals.h

```diff
@@ -419,11 +440,22 @@ PYBIND11_NOINLINE internals &get_internals() {
     } gil;
     error_scope err_scope;
 
-    PYBIND11_STR_TYPE id(PYBIND11_INTERNALS_ID);
-    auto builtins = handle(PyEval_GetBuiltins());
-    if (builtins.contains(id) && isinstance<capsule>(builtins[id])) {
-        internals_pp = static_cast<internals **>(capsule(builtins[id]));
+    constexpr const char *id_cstr = PYBIND11_INTERNALS_ID;
+    str id(id_cstr);
+
+    dict state_dict = get_internals_state_dict();
+
+    if (state_dict.contains(id_cstr)) {
+        object o = state_dict[id];
+        // May fail if 'capsule_obj' is not a capsule, or if it has a different
+        // name. We clear the error status below in that case
+        internals_pp = static_cast<internals **>(PyCapsule_GetPointer(o.ptr(), id_cstr));
```

Comment:
Why the raw CAPI here? Any particular reason we are changing it from pytype.h API it was before?

Reply:
This is my current version:

```
        void *raw_ptr = PyCapsule_GetPointer(state_dict[id].ptr(), id_cstr);
        if (raw_ptr == nullptr) {
            raise_from(
                PyExc_SystemError,
                "pybind11::detail::get_internals(): Retrieve internals** from capsule FAILED");
        }
        internals_pp = static_cast<internals **>(raw_ptr);
```

The nice thing is that `PyCapsule_GetPointer()` does everything "just right" in one simple line. Additionally, in this particular situation I'd definitely want to avoid the `throw PYBIND11_OBJECT_CHECK_FAILED()` that comes with the `capsule` constructor.

I don't have a good idea for using `capsule` in an elegant way here TBH.

Reply:
While looking around more (work under PR #4329) I noticed this code in embed.h (`finalize_interpreter()`):

```
    handle builtins(PyEval_GetBuiltins());
    ...
    if (builtins.contains(id) && isinstance<capsule>(builtins[id])) {
        internals_ptr_ptr = capsule(builtins[id]);
    }
```

Two things learned:

* That needs to be changed, too, to inspect the correct dict.
* The implementation is a bit on the high-level overkill side, scraping by an `if (...) throw`, but it is elegant!